src/toast/ops/polyfilter.py

```diff
@@ -0,0 +1,607 @@
+from ..utils import Logger, Environment, Timer, GlobalTimers, dtype_to_aligned
+
+
+from .._libtoast import filter_polynomial
+
+
+
+
+XAXIS, YAXIS, ZAXIS = np.eye(3)
+
+
+@trait_docs
+class PolyFilter2D(Operator):
+    """Operator to regress out 2D polynomials across the focal plane."""
+
+    API = Int(0, help="Internal interface version for this operator")
+
+    det_data = Unicode("signal", help="Observation detdata key apply the gain error to")
+
+    pattern = Unicode(
+        f".*",
+        allow_none=True,
+        help="Regex pattern to match against detector names. Only detectors that "
+        "match the pattern are filtered.",
+    )
+
+    order = Int(1, allow_none=False, help="Polynomial order")
+
+    det_flags = Unicode(
+        None, allow_none=True, help="Observation detdata key for flags to use"
+    )
+
+    det_flag_mask = Int(1, help="Bit mask value for optional detector flagging")
+
+    poly_flag_mask = Int(1, help="Bit mask value for intervals that fail to filter")
+
+    shared_flags = Unicode(
+        None, allow_none=True, help="Observation shared key for telescope flags to use"
+    )
+
+    shared_flag_mask = Int(1, help="Bit mask value for optional shared flagging")
+
+    view = Unicode(
+        None, allow_none=True, help="Use this view of the data in all observations"
+    )
+
+    focalplane_key = Unicode(None, allow_none=True, help="Which focalplane key to match")
+
+    @traitlets.validate("shared_flag_mask")
+    def _check_shared_flag_mask(self, proposal):
+        check = proposal["value"]
+        if check < 0:
+            raise traitlets.TraitError("Shared flag mask should be a positive integer")
+        return check
+
+    @traitlets.validate("det_flag_mask")
+    def _check_det_flag_mask(self, proposal):
+        check = proposal["value"]
+        if check < 0:
+            raise traitlets.TraitError("Det flag mask should be a positive integer")
+        return check
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        return
+
+    @function_timer
+    def _exec(self, data, detectors=None, **kwargs):
+        """Apply the 2D polynomial filter to the signal.
+
+        Args:
+            data (toast.Data): The distributed data.
+
+        """
+        if detectors is not None:
+            raise RuntimeError("PolyFilter2D cannot be run in batch mode")
+        norder = self.order + 1
+        nmode = norder ** 2
+        pat = re.compile(self.pattern)
+
+        for obs in data.obs:
+            t0 = time()
+            t_template = 0
+            t_get_norm = 0
+            t_apply_norm = 0
+            t_solve = 0
+            t_clean = 0
+
+            # communicator for processes with the same sample range
+            comm = obs.comm_col
+
+            # Detectors to process
+
+            detectors = []
+            for det in obs.all_detectors:
+                if pat.match(det) is None:
```

Comment:
The `Pipeline` operator can be used to execute multiple sub-operators one "detector set" at a time (not just one detector at a time).  However, we have explicitly banned that mode of operation above (we don't allow calling `exec()` with a subset of detectors).  So if we want to filter det sets (as defined in each observation), we would need to use some other method of selecting each detector set (either a focalplane key or a regex that matches their names).  I guess this is fine for now, but at some point we might want to add extra logic to check that if a subset of detectors is given, *AND* if it is a detector set, then allow it.